### PR TITLE
Add abstract base class for attention mechanisms with unified interface

### PR DESCRIPTION
**Pull request by iseeyuan**
### Summary
Add abstract base class for attention mechanisms with unified interface.
It creates the interface to multiple Attention definitions, like NPU friendly attentions, or other attention types like Multi-Head Latent Attention (MLA) used in Deepseek.
A simple registry is provided to easily add and register a new attention class.

Moved the current attention implementation to attention.py and rename it to AttentionMHA. 

### Test plan
CI

### CODE REVIEW

File: examples/models/llama/attention.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    @abstractmethod
+    def forward(
+        self,
+        x: torch.Tensor,
+        freqs_cos: torch.Tensor,
+        freqs_sin: torch.Tensor,
+        mask: Optional[torch.Tensor] = None,
+        input_pos: Optional[torch.Tensor] = None,
+        in_cache_state: Optional[Any] = None,
+        out_cache_state: Optional[Any] = None,
```
Comment (iseeyuan) [inline examples/models/llama/attention.py:19]: replace them with kwargs

**Comment by iseeyuan** [comment]
@CypherpunkSamurai I'm trying to complete the refactor and land it this week. Let me create an issue of adding the MLA to this interface.

File: examples/models/llama/attention.py (hunk 1 above)
Comment (iseeyuan) [reply examples/models/llama/attention.py:19]: directly using kwargs may break type-safe. Keep them as is and consider using TypedDict and Unpack for kwarg type checking later.

**Review by jackzhxng — COMMENTED** [review]
Is this ready for review? Also would probably tag with an actual release note tag since I assume this would be good to highlight in 0.6 release notes

File: examples/models/llama/attention.py [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+from typing import Any, Dict, Optional, Tuple, Type
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from executorch.examples.models.llama.model_args import ModelArgs
+from executorch.examples.models.llama.rope import Rope
+
+
+class Attention(nn.Module, ABC):
```
Comment (sxu) [inline examples/models/llama/attention.py:11]: So far a specialized implementation is only used during lowering and on device, and it needs to be able to accept checkpoint from whatever definition was used during training. What do see as the usage pattern going forward? Is the `AttentionMHA` below the standard definition that specialization of this class needs to support?

Comment (iseeyuan) [reply examples/models/llama/attention.py:11]: @sxu 
> Is the AttentionMHA below the standard definition that specialization of this class needs to support?
Not necessarily. The attention type is added into the model args. Usually the model arg and checkpoint will be saved in one place. We use model arg to build the model, and load the checkpoint as state_dict. If the checkpoint does not match the model architecture there will be error. We don't break the standard process of PyTorch.

**Comment by iseeyuan** [comment]
> Is this ready for review? Also would probably tag with an actual release note tag since I assume this would be good to highlight in 0.6 release notes

@dvorjackz It's currently in the example model. I'm fine to promote this to extension/llm, rename it to llm_transformer, or simply transformer, and mark it user faced.

File: examples/models/llama/attention.py (hunk 2 above)
Comment (sxu) [reply examples/models/llama/attention.py:11]: I see, we don't usual expect the training to be done on a specialized NPU implementation, but I guess we can tweak the state dict loading on a case by case basis.

**Review by sxu — COMMENTED** [review]
Looks reasonable to me. The cache, mask, and rope frequencies still need to passed down from the top level transformer and updates from each layer returned need to be returned in a follow up PR.

**Review by tarun292 — APPROVED** [review]

**Comment by iseeyuan** [comment]
> Looks reasonable to me. The cache, mask, and rope frequencies still need to passed down from the top level transformer and updates from each layer returned need to be returned in a follow up PR.

Thanks @sxu ! I'll land this one when all CI pass, and please feel free to add further PRs to improve it.
